src/os/bluestore/NVMEDevice.cc

```diff
@@ -48,14 +45,14 @@
 #undef dout_prefix
 #define dout_prefix *_dout << "bdev(" << sn << ") "
 
-static constexpr uint16_t data_buffer_default_num = 2048;
+thread_local SharedDriverQueueData *queue_t;
```

Comment:
@optimistyzy @yuyuyu101 

as reported by @howard0su, this causes a leak. as each thread which `NVMEDevice::aio_submit()` with pending io will create a new `queue_t` if it's not available in this thread yet. but only the thread which calls `NVMEDevice::close()` will destruct the `queue_t` instance once for that thread.

also, in his case the number of qpair is running out on a 20 core host, with following error message

```
nvme_ctrlr.c: 308:spdk_nvme_ctrlr_alloc_io_qpair: *ERROR*: No free I/O queue IDs
```

Reply:
I only see the main thread call the destructor which only deletes one queue object. other objects in the other threads are still there. and those instances are reference a delete NVMEDevice object. If the thread is still using or reuse some time later, it will crash as well.

In general, it is bad idea to putting a object reference in TLS.